### PR TITLE
Use Simulacrum for F[_, _] type classes

### PR DESCRIPTION
**Pull request by ceedubs**
This addresses #1064. I updated all of the type classes that I could find that I think Simulacrum 0.8 supports, but I'm not 100% I got them all. There was some discussion on that issue about moving arrow into its own package, but I think we should treat that as a separate concern.

~~I currently have this marked as `Do not merge`, because it depends on a snapshot version of Simulacrum. @mpilquist as far as I can tell this feature is working splendidly. Do you think Simulacrum is ready for an official 0.8 release?~~

@yilinwei I know you had started some work on this. I hope I'm not stepping on your toes. I wanted to try to wrap this up in preparation for cats 1.0.

### CODE REVIEW

**Comment by ceedubs** [comment]
@mpilquist I was under the assumption that Simulacrum still doesn't support shapes like `ApplicativeError`. Is that right?

**Comment by yilinwei** [comment]
@ceedubs Nope, not yet.

**Comment by ceedubs** [comment]
Version 0.8.0 of Simulacrum has been released now. When I'm at the right computer I just need to update this PR to depend on it.

**Comment by ceedubs** [comment]
I've updated this to Simulacrum 0.8.0. It should be ready for review/merge.

**Comment by travisbrown** [comment]
I haven't really been following Simulacrum development closely enough to have anything useful to say here, but this looks reasonable and it works, so we might as well update before 0.7.0.

**Comment by ceedubs** [comment]
I've resolved the merge conflict. @mpilquist does this look good to you?

File: core/src/main/scala/cats/Bifoldable.scala [hunk 1]
```diff
@@ -1,9 +1,11 @@
 package cats
 
+import simulacrum.typeclass
+
 /**
  * A type class abstracting over types that give rise to two independent [[cats.Foldable]]s.
  */
-trait Bifoldable[F[_, _]] extends Any with Serializable { self =>
+@typeclass trait Bifoldable[F[_, _]] extends Any with Serializable { self =>
```
Comment (mpilquist) [inline core/src/main/scala/cats/Bifoldable.scala]: No need for `Any with Serializable` here -- Simulacrum adds that.

File: core/src/main/scala/cats/functor/Bifunctor.scala [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 package functor
 
+import simulacrum.typeclass
+
 /**
  * A type class of types which give rise to two independent, covariant
  * functors.
  */
-trait Bifunctor[F[_, _]] extends Any with Serializable { self =>
+@typeclass trait Bifunctor[F[_, _]] extends Any with Serializable { self =>
```
Comment (mpilquist) [inline core/src/main/scala/cats/functor/Bifunctor.scala]: Same as above

File: core/src/main/scala/cats/functor/Profunctor.scala [hunk 3]
```diff
@@ -1,15 +1,30 @@
...
+import simulacrum.typeclass
+
 /**
  * A [[Profunctor]] is a [[Contravariant]] functor on its first type parameter
  * and a [[Functor]] on its second type parameter.
  *
  * Must obey the laws defined in cats.laws.ProfunctorLaws.
  */
-trait Profunctor[F[_, _]] extends Serializable { self =>
+@typeclass trait Profunctor[F[_, _]] extends Serializable { self =>
```
Comment (mpilquist) [inline core/src/main/scala/cats/functor/Profunctor.scala]: No need for Serializable

**Comment by mpilquist** [comment]
I added a couple of minor comments but 👍 otherwise.

**Comment by ceedubs** [comment]
Thanks @mpilquist. I updated the ones that you pointed out (and also `Compose`).

**Pull request merged**